apps/voting/app/src/App.js

```diff
@@ -1,352 +1,276 @@
+          <br />
+        </React.Fragment>
+      ))}
+    </AutoLink>
+  ) : null
+}
 
-  constructor(props) {
-    super(props)
+// Generate a cache ID from a list of votes, based on their voteId.
+const cacheIdFromVotes = votes =>
+  votes
+    ? votes
+        .map(vote => vote.voteId)
+        .sort()
+        .join()
+    : ''
 
-    this.state = {
-      createVoteVisible: false,
-      currentVoteId: -1,
-      tokenContract: this.getTokenContract(props.tokenAddress),
-      voteVisible: false,
-      voteSidebarOpened: false,
-      userAccountVotes: new Map(),
-    }
-  }
-  componentWillReceiveProps(nextProps) {
-    // Refresh the token contract if its address changes
-    if (nextProps.tokenAddress !== this.props.tokenAddress) {
-      this.setState({
-        tokenContract: this.getTokenContract(nextProps.tokenAddress),
-      })
-    }
+function App() {
+  const { api, network, appState, connectedAccount } = useAragonApi()
 
-    // Refresh the account votes if the account changes,
-    // or if there is any vote update.
-    if (
-      nextProps.votes !== this.props.votes ||
-      nextProps.userAccount !== this.props.userAccount
-    ) {
-      this.loadUserAccountVotes(nextProps.userAccount, nextProps.votes)
-    }
-  }
+  const [createVoteVisible, setCreateVoteVisible] = useState(false)
+  const [currentVoteId, setCurrentVoteId] = useState(-1)
+  const [tokenContract, setTokenContract] = useState(null)
+  const [voteVisible, setVoteVisible] = useState(false)
+  const [voteSidebarOpened, setVoteSidebarOpened] = useState(false)
+  const [userAccountVotes, setUserAccountVotes] = useState(new Map())
+  const now = useNow()
 
-  async loadUserAccountVotes(userAccount, votes) {
-    const { app } = this.props
+  const {
+    appStateReady,
+    pctBase,
+    tokenDecimals,
+    tokenSymbol,
+    voteTime,
+  } = appState
 
-    if (!userAccount) {
-      this.setState({ userAccountVotes: new Map() })
-      return
-    }
+  // Add some useful data to render the votes
+  const votes = useMemo(
+    () =>
+      appStateReady
+        ? appState.votes.map(vote => ({
+            ...vote,
+            data: {
+              ...vote.data,
+              open: isVoteOpen(vote, now),
 
-    this.setState({
-      userAccountVotes: new Map(
-        await Promise.all(
-          votes.map(vote =>
-            app
-              .call('getVoterState', vote.voteId, userAccount)
-              .toPromise()
-              .then(result => [vote.voteId, result])
-          )
-        )
-      ),
-    })
-  }
+              // Render text fields
+              descriptionNode: voteTextNode(vote.data.description),
+              metadataNode: voteTextNode(vote.data.metadata),
+            },
+            userAccountVote: voteTypeFromContractEnum(
+              userAccountVotes.get(vote.voteId)
+            ),
+          }))
+        : appState.votes,
+    [appStateReady, cacheIdFromVotes(appState.votes), userAccountVotes]
```

Comment:
⚠️ I think this caching for this effect is too aggressive and will result in confusing results.

Using `cacheIdFromVotes()` means `votes` will only get recomputed if a new vote is added. **But** it will not be recomputed if an existing vote's data was modified (e.g. its voting values change). Nor would it recompute whether the vote was still open or not, since `now`'s not in the dependencies list.

Perhaps we can instead apply caching for the rendered metadata and description nodes using a `Map`?

Reply:
I removed the caching for now, let’s add a throttling mechanism to `@aragon/api` / `@aragon/api-react` as discussed offline :+1: